engine/notifier_test.go

```diff
@@ -83,54 +83,62 @@ func TestNotifier_ManyNotifications(t *testing.T) {
 // sends just as many notifications with small delays. We require that
 // all workers eventually get a notification.
 func TestNotifier_ManyConsumers(t *testing.T) {
-	t.Parallel()
-	notifier := NewNotifier()
-	c := notifier.Channel()
+	singleTestRun := func(t *testing.T) {
+		t.Parallel()
+		notifier := NewNotifier()
+		c := notifier.Channel()
 
-	// spawn 100 worker routines to each wait for a notification
-	var startingWorkers sync.WaitGroup
-	pendingWorkers := atomic.NewInt32(100)
-	for i := 0; i < 100; i++ {
-		startingWorkers.Add(1)
-		go func() {
-			startingWorkers.Done()
-			<-c
-			pendingWorkers.Dec()
-		}()
-	}
-	startingWorkers.Wait()
-	time.Sleep(10 * time.Millisecond)
+		// spawn 100 worker routines to each wait for a notification
+		var startingWorkers sync.WaitGroup
+		pendingWorkers := atomic.NewInt32(100)
+		for i := 0; i < 100; i++ {
+			startingWorkers.Add(1)
+			go func() {
+				startingWorkers.Done()
+				<-c
+				pendingWorkers.Dec()
+			}()
+		}
+		startingWorkers.Wait()
 
-	// send 100 notifications, with small delays
-	for i := 0; i < 100; i++ {
-		notifier.Notify()
-		time.Sleep(10 * time.Microsecond)
+		// send 100 notifications, with small delays
+		for i := 0; i < 100; i++ {
+			notifier.Notify()
+		}
```

Comment:
This is probably unlikely, but I do see a race condition here *in theory*. Let's reduce the example to 2 workers instead of 100 for simplicity. 

Consider the following execution order for the 3 threads (the main thread, and the 2 worker routines):
* worker 1 executes line 97
* worker 2 executes line 97
* main thread returns from call on line 102
* main thread executes line 105 (first loop iteration)
* main thread executes line 106 (first loop iteration)
* main thread executes line 105 (second loop iteration)
* main thread executes line 106 (second loop iteration) *NOTIFICATION DROPPED*
* worker 1 executes line 98 (returns immediately)
* worker 2 executes line 98 (blocks forever)

In short, the loop in the main thread sends *both* notifications before either worker routine has a chance to read from the notification channel, which results in the second routine never being notified.

All this to say, perhaps it would be good to keep the `time.Sleep()` after each loop iteration?

Reply:
great catch. totally agree.